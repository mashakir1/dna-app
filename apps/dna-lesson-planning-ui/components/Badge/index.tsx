// Styles
import styles from "./badge.module.scss";

export const Badge: ({
  text,
  classNames,
}: {
  text: string;
  classNames?: string;
}) => JSX.Element = ({ text, classNames }) => (
  <div className={`${styles.badge} ${classNames} button`}>
    <span>{text}</span>
  </div>
);
