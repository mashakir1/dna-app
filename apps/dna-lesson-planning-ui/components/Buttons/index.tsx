export { IconLinkButton } from "./IconLinkButton";
export { OutlineLinkButton } from "./OutlineLinkButton";
export { FilledButton } from "./FilledButton";
export { FilledLinkButton } from "./FilledLinkButton";
export { EmptyButton } from "./EmptyButton";
export { EmptyIconButton } from "./EmptyIconButton";
