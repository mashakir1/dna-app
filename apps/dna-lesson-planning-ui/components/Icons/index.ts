export { ClickableIcon } from "./ClickableIcon";
export { IconWithBackground } from "./IconWithBackground";
