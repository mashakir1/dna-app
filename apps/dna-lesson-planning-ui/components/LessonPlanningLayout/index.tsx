// Imports
import { ReactNode } from "react";

// Components
import { Header } from "./Header";

// Styles
import styles from "./LessonPlanningLayout.module.scss";

const Layout = ({ children }: { children: ReactNode }) => (
  <div className={styles.page}>
    <Header />
    <div className={`${styles.container} flex column gap-l`}>{children}</div>
  </div>
);
export default Layout;
