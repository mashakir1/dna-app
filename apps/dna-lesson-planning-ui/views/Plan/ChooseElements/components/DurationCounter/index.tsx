// Imports
import { useMemo } from "react";
import useTranslation from "next-translate/useTranslation";

// Components
import { ClockIcon } from "components/Svg";

// Styles
import styles from "./durationCounter.module.scss";

export interface DurationCounterProps {
  minutes: number;
}

const DurationCounter: (props: DurationCounterProps) => JSX.Element = ({ minutes }) => {
  const { t } = useTranslation("lessonPlanning");
  const { hours, durationUnit, durationValue } = useMemo(() => {
    const hours = (minutes - (minutes % 60)) / 60;
    return {
      hours,
      durationUnit: hours > 0 ? t("hour-symbol") : t("minute-symbol"),
      durationValue: hours > 0 ? hours : minutes
    };
  }, [t, minutes]);

  return <div className={`${styles.durationCounter} flex column justify-center gap-xs items-center`}>
    <div className="flex gap-xs items-center">
      <p className="large"><b>{durationValue}</b></p>
      <p className="small">{durationUnit}</p>
      {hours > 0 && minutes % 60 > 0 && <p className="large"><b>{minutes % 60}</b></p>}
    </div>
    <ClockIcon />
  </div>;
};

export default DurationCounter;
