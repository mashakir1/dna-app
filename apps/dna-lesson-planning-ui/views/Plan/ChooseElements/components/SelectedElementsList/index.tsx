// Imports
import { Dispatch, SetStateAction } from "react";

// Components
import Dropbox from "../Dropbox";

// Types
import { Element } from "types";

// Styles
import styles from "./selectedElementsList.module.scss";

export interface SelectedElementsListProps {
  selectedIds: Array<string>;
  setSelectedIds: Dispatch<SetStateAction<Array<string>>>;
  selectedElements: Array<Element>;
  elementCrossOnClick: (id: string) => void;
}

const SelectedElementsList: (props: SelectedElementsListProps) => JSX.Element = ({
  selectedIds, setSelectedIds, selectedElements, elementCrossOnClick
}) => {
  return <div className={`${styles.selectedElementsList} flex gap-s`}>
    {selectedElements?.map((element, index) => <Dropbox
      key={element.id}
      id={index}
      selectedIds={selectedIds}
      setSelectedIds={setSelectedIds}
      element={element}
      elementCrossOnClick={elementCrossOnClick}
    />)}
  </div>;
};

export default SelectedElementsList;
