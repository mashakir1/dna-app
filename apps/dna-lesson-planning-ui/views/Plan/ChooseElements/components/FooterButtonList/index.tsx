// Imports
import useTranslation from "next-translate/useTranslation";

// Components
import { FilledLinkButton, OutlineLinkButton } from "components/Buttons";

// Styles
import styles from "./footerButtonList.module.scss";

export interface FooterButtonListProps {
  isClickable: boolean;
}

const FooterButtonList: (props: FooterButtonListProps) => JSX.Element = ({ isClickable }) => {
  const { t } = useTranslation("lessonPlanning");
  return <div className={`${styles.footerButtonList} flex column gap-s justify-center`}>
    <FilledLinkButton text={t("continue")} href={"/plan/review"} disabled={!isClickable} />
    <OutlineLinkButton text={t("edit-slides")} href={"/plan/subject"} disabled={!isClickable} />
  </div>;
};

export default FooterButtonList;
