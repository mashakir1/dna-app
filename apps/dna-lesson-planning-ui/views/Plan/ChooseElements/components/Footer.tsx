// Imports
import { Dispatch, SetStateAction } from "react";

// Components
import DurationCounter from "./DurationCounter";
import FooterButtonList from "./FooterButtonList";
import SelectedElementsList from "./SelectedElementsList";

// Types
import { Element } from "types";

// Styles
import styles from "./footer.module.scss";

interface FooterProps {
  selectedIds: Array<string>;
  setSelectedIds: Dispatch<SetStateAction<Array<string>>>;
  elements: Array<Element>;
}

export const Footer: (props: FooterProps) => JSX.Element = ({
  selectedIds, setSelectedIds, elements
}) => {
  const selectedElements = selectedIds.map(
    id => elements.find(element => element.id === id)
  );

  const totalDuration = selectedElements
    ?.map(element => element.duration)
    ?.reduce((a, b) => a + b, 0);

  const elementCrossOnClick = (selectedElementId: string) => {
    setSelectedIds(selectedIds.filter(id => id !== selectedElementId));
  };

  return <div className={`${styles.footer} flex gap-m items-center`}>
    <DurationCounter minutes={totalDuration} />
    <SelectedElementsList
      selectedIds={selectedIds}
      setSelectedIds={setSelectedIds}
      selectedElements={selectedElements}
      elementCrossOnClick={elementCrossOnClick}
    />
    <FooterButtonList isClickable={selectedElements?.length > 0} />
  </div>;
};

