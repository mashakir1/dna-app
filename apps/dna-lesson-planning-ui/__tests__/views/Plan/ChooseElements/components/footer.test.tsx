import { render, screen } from "@testing-library/react";
import { Footer } from "views/Plan/ChooseElements/components/Footer";
import { testElement } from "../mockData";
import { DurationCounterProps } from "views/Plan/ChooseElements/components/DurationCounter";
import { SelectedElementsListProps } from "views/Plan/ChooseElements/components/SelectedElementsList";
import { FooterButtonListProps } from "views/Plan/ChooseElements/components/FooterButtonList";

const durationCounterMock = (props: DurationCounterProps) => <div data-testid="DurationCounter">{props.minutes}</div>;
jest.mock(
  "views/Plan/ChooseElements/components/DurationCounter",
  () => durationCounterMock
);

const selectedElementListMock = (props: SelectedElementsListProps) => <div data-testid="selectedElementList">{
  props.selectedElements.map(element => <div data-testid={element.id} key={element.id} />)
}</div>;
jest.mock(
  "views/Plan/ChooseElements/components/SelectedElementsList",
  () => selectedElementListMock
);

const footerButtonListMock = (props: FooterButtonListProps) => <div data-testid="FooterButtonList">{
  props.isClickable ? "Clickable" : "Not clickable"
}</div>;
jest.mock(
  "views/Plan/ChooseElements/components/FooterButtonList",
  () => footerButtonListMock
);

describe("Footer", () => {
  it("matches empty snapshot", () => {
    const { container } = render(<Footer
      elements={[]} selectedIds={[]} setSelectedIds={() => undefined}
    />);

    expect(container).toMatchSnapshot();
  });

  it("matches populated snapshot", () => {
    const firstElement = { ...testElement, id: "first" };
    const secondElement = { ...testElement, id: "second" };

    const { container } = render(<Footer
      elements={[firstElement, secondElement]}
      selectedIds={["first", "second"]}
      setSelectedIds={() => undefined}
    />);

    expect(container).toMatchSnapshot();
  });

  it("passes selected Elements to SelectedElementsList correctly", () => {
    const firstElement = { ...testElement, id: "success" };
    const secondElement = { ...testElement, id: "fail" };

    render(<Footer
      elements={[firstElement, secondElement]}
      selectedIds={["success"]}
      setSelectedIds={() => undefined}
    />);

    expect(screen.queryByTestId("success")).toBeTruthy();
    expect(screen.queryByTestId("fail")).toBeFalsy();
  });

  it("calculates duration correctly", () => {
    const firstElement = { ...testElement, id: "first", duration: 20 };
    const secondElement = { ...testElement, id: "second", duration: 30 };

    render(<Footer
      elements={[firstElement, secondElement]}
      selectedIds={["first", "second"]}
      setSelectedIds={() => undefined}
    />);

    expect(screen.queryByText("50")).toBeTruthy();
  });

  it("buttons disabled if no elements selected", () => {
    const firstElement = { ...testElement, id: "first", duration: 20 };
    const secondElement = { ...testElement, id: "second", duration: 30 };

    render(<Footer
      elements={[firstElement, secondElement]}
      selectedIds={[]}
      setSelectedIds={() => undefined}
    />);

    expect(screen.queryByText("Not clickable")).toBeTruthy();
  });

  it("buttons enabled if at least one element selected", () => {
    const firstElement = { ...testElement, id: "first", duration: 20 };
    const secondElement = { ...testElement, id: "second", duration: 30 };

    render(<Footer
      elements={[firstElement, secondElement]}
      selectedIds={["first"]}
      setSelectedIds={() => undefined}
    />);

    expect(screen.queryByText("Clickable")).toBeTruthy();
  });
});
