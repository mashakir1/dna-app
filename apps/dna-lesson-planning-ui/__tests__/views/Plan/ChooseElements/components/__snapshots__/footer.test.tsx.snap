// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer matches empty snapshot 1`] = `
<div>
  <div
    class="footer flex gap-m items-center"
  >
    <div
      data-testid="DurationCounter"
    >
      0
    </div>
    <div
      data-testid="selectedElementList"
    />
    <div
      data-testid="FooterButtonList"
    >
      Not clickable
    </div>
  </div>
</div>
`;

exports[`Footer matches populated snapshot 1`] = `
<div>
  <div
    class="footer flex gap-m items-center"
  >
    <div
      data-testid="DurationCounter"
    >
      0
    </div>
    <div
      data-testid="selectedElementList"
    >
      <div
        data-testid="first"
      />
      <div
        data-testid="second"
      />
    </div>
    <div
      data-testid="FooterButtonList"
    >
      Clickable
    </div>
  </div>
</div>
`;
